app/jobs/shipit/update_github_last_deployed_ref_job.rb

```diff
@@ -0,0 +1,37 @@
+module Shipit
+  class UpdateGithubLastDeployedRefJob < BackgroundJob
+    queue_as :default
+
+    DEPLOY_PREFIX = 'shipit-deploy'.freeze
+
+    def perform(stack)
+      stack_sha = stack.last_deployed_commit.sha
+      environment = stack.environment
+      stack_ref = create_full_ref(environment)
+      client = Shipit.github.api
+
+      full_repo_name = stack.github_repo_name
+
+      update_or_create_ref(client: client, repo_name: full_repo_name, ref: stack_ref, new_sha: stack_sha)
+    end
+
+    private
+
+    def create_full_ref(stack_environment)
+      [DEPLOY_PREFIX, stack_environment].join("/")
+    end
+
+    def create_ref(client:, repo_name:, ref:, sha:)
+      client.create_ref(repo_name, ref, sha)
+    end
+
+    def update_or_create_ref(client:, repo_name:, ref:, new_sha:)
+      client.update_ref(repo_name, ref, new_sha)
```

Comment:
When the request is valid (e.g. the ref exists), updating is Idempotent, to the extent that repeated Update calls will return 200 and not affect the ref (the SHA remains the same).
A 422 with a specific message is returned when the ref doesn't exist, so only in that case do we create the ref.

The most common case in Shipit will be updating a ref - creation will usually only happen for a completely new stack. Therefore, this is much more efficient overall than trying to check for existence in advance (i.e. by switching off of `GET /ref/`)